### DISCUSSION TITLE
How to read an int from stdin?

### DISCUSSION THREAD

**Discussion by lifubang** [denoland/deno#15872] [Q&A] [answered]
```
acmcoder@MacBook-Pro deno % cat readInt.ts 
import { BufReader } from "https://deno.land/std@0.155.0/io/buffer.ts";
import { readInt } from "https://deno.land/std@0.155.0/io/ioutil.ts";
let reader = new BufReader(Deno.stdin);
let abc = await readInt(reader);
console.log(abc)
acmcoder@MacBook-Pro deno % deno run readInt.ts
123 456
825373472
```

**Comment by ayame113** [comment, marked as answer]
I'm not sure if I understand the problem correctly, but what about something like this?

```ts
const reader = Deno.stdin.readable.pipeThrough(new TextDecoderStream());
for await (const chunk of reader) {
  console.log(Number(chunk));
}
```

```shell
> deno run readInt.ts
123
123
456
456
```

**Reply by lifubang** [reply]
Yes, it is. But there is another problem, we can't run it correctly using linux pipe, for example:
```
acmcoder@MacBook-Pro deno % cat 1.in
123
456
acmcoder@MacBook-Pro deno % cat 1.in | deno run testInt.ts
NaN
```

**Reply by lifubang** [reply]
> I'm not sure if I understand the problem correctly

I want to write some code in `node.js` or `deno` to calc A+B from stdin, but it is very difficulty.
The `C` code is very simple:
```c
#include <stdio.h>
int main()
{
  int a, b;
  while(scanf("%d %d", &a, &b) != EOF)
    printf("%d\n", a + b);
}
```

**Reply by lifubang** [reply]
The test case may be like this:
```
123 456
789
123
```
We will run it like this: `cat 1.in | ./Main`, the output will be:
```
579
912
```

**Reply by ayame113** [reply]
> I want to write some code in node.js or deno to calc A+B from stdin, but it is very difficulty.

It's interesting!! Deno doesn't have anything like scanf, so you'll have to control the newline and space delimiters yourself.
this works for me:

```ts
import { TextLineStream } from "https://deno.land/std@0.155.0/streams/mod.ts";

const lines = Deno.stdin.readable
  .pipeThrough(new TextDecoderStream()) // convert byte array to string
  .pipeThrough(new TextLineStream()); // make the stream newline delimited

let i = 0;
let prev: string | undefined;
for await (const line of lines) {
  for (const num of line.split(" ")) {
    if (i % 2) {
      console.log(Number(num) + Number(prev));
    }
    prev = num;
    i++;
  }
}
```

**Comment by lifubang** [comment]
> ```
> acmcoder@MacBook-Pro deno % cat readInt.ts 
> import { BufReader } from "https://deno.land/std@0.155.0/io/buffer.ts";
> import { readInt } from "https://deno.land/std@0.155.0/io/ioutil.ts";
> let reader = new BufReader(Deno.stdin);
> let abc = await readInt(reader);
> console.log(abc)
> acmcoder@MacBook-Pro deno % deno run readInt.ts
> 123 456
> 825373472
> ```

@ayame113 Could I ask one more question? Why `readInt` can't be used in here?

**Reply by 0f-0b** [reply]
`readInt` is for reading binary data, similar to `fread` in C.

```shellsession
$ cat main.c
#include <arpa/inet.h>
#include <inttypes.h>
#include <stdio.h>

int main(void) {
  int32_t abc;
  fread(&abc, sizeof abc, 1, stdin);
  printf("%" PRId32 "\n", ntohl(abc));
}

$ cc main.c && ./a.out
123 456
825373472
```
